text/0000-return-type-notation.md

```diff
@@ -0,0 +1,1105 @@
+
+Finally, while `typeof` clearly is a more general feature, it's not clear how well motivated that generality is. The main use cases we have in mind are more naturally and directly handled by RTN. To justify `typeof`, we'd want to have a solid rationale of use cases.
+
+## Why not make *all* futures `Send`?
+
+The `#[async_trait]` macro solves the send bounds problem by forcing the trait to declare up front whether it will require send or not. This is required by the desugaring that async-trait uses. For many users, this is a fine solution, since they always work with sendable futures. But there are a significant set of users that do not want send bounds, either because they are in an embedded context or because they are using a thread-per-core architecture. The widely used tokio runtime, for example, can be configured to either use work-stealing (which requires `Send` futures) or to be a single-threaded executor (which does not). The `glommio` executor does not require `Send` bounds on futures because it never moves tasks between threads. The Fuchsia project makes extensive use of single-threaded executors in their runtime, and hence they do not require `Send` bounds. The `embassy` runtime targets embedded environments that only have a uniprocessor and which have no need for `Send` bounds. All of these environments are disadvantaged by defaults that require send bounds.
+
+One of our design goals with async-trait is to support core interoperability traits for things like reading, writing, HTTP, etc. The whole point of these traits is to be usable across many runtimes. If those traits forced `Send` bounds, that would be unnecessarily limiting, which would lead to users of non-Send-requiring runtimes to avoid them. If the traits did NOT force `Send` bounds, they would not be compatible with work stealing runtimes (the most popular choice) unless there was some additional feature to "opt-in" to needing send bounds -- which is exactly the gap RTN is looking to close.
+
+## Why not create an associated type that represents the return type?
+
+Early on in our design work, we expected to simply create an associated type within the trait to represent the return type. For example this trait:
+
+```rust
+trait Factory {
+    fn widgets(&self) -> impl Iterator<Item = Widget>;
+}
+```
+
+might have been desugared as follows:
+
+```rust
+trait Factory {
+    type widgets<'a>: Iterator<Item = Widget>; // 👈 implicitly introduced
+    fn widgets(&self) -> Self::widgets<'_>;
+}
+```
+
+This would mean that users could write a bound on `F::widgets` to bound the return type of `widgets`
+
+```rust
+fn use_factory<F>()
+where
+    F: Factory,
+    for<'a> F::widgets<'a>: Send,
+{}
+```
+
+We encountered a number of problems with this design.
+
+### If the name is implicit, what name should we use?
```

Comment:
you could also use one of those `#` namespace such as `fn#widgets` or `return#widgets` or `result_of#widgets`

Reply:
yes, I'll add that as a possibility